src/App/mermaidData/pullApiData.js

```diff
@@ -44,80 +62,111 @@ export const pullApiData = async ({
-            // we still delete project related data in addition to anything in the removes array,
-            // because the removes array response doesnt include all the things we want to delete currently.
-            const deleteProjectRelatedRecords = dexiePerUserDataInstance[apiDataType]
-              .where({ project: projectId })
-              .delete()
-            const resetProjectRelatedLastRevisionNumbers =
-              resetLastRevisionNumberForProjectDataType({
-                dataType: apiDataType,
-                projectId,
-                dexiePerUserDataInstance,
-              })
-
-            const removeOfflineReadyProjectStatus =
-              dexiePerUserDataInstance.uiState_offlineReadyProjects.delete(projectId)
-
-            await Promise.all([
-              deleteProjectRelatedRecords,
-              resetProjectRelatedLastRevisionNumbers,
-              removeOfflineReadyProjectStatus,
-            ])
+  const _evaluatePullResponseAndUpdateOfflineMermaidDataAccordingly =
+    await dexiePerUserDataInstance.transaction(
+      'rw',
+      dexiePerUserDataInstance.benthic_attributes,
+      dexiePerUserDataInstance.choices,
+      dexiePerUserDataInstance.collect_records,
+      dexiePerUserDataInstance.fish_families,
+      dexiePerUserDataInstance.fish_genera,
+      dexiePerUserDataInstance.fish_species,
+      dexiePerUserDataInstance.project_managements,
+      dexiePerUserDataInstance.project_profiles,
+      dexiePerUserDataInstance.project_sites,
+      dexiePerUserDataInstance.projects,
+      dexiePerUserDataInstance.uiState_lastRevisionNumbersPulled,
+      dexiePerUserDataInstance.uiState_offlineReadyProjects,
+      async () => {
+        await persistLastRevisionNumbersPulled({
+          dexiePerUserDataInstance,
+          apiData,
+          projectId,
+        })
+
+        apiDataNamesToPull.forEach(async (apiDataType) => {
+          if (apiDataType === 'choices') {
+            // choices deletes property will always be empty, so we just ignore it
+            // additionally the updates property is an object, not an array, so we just store it directly
+
+            dexiePerUserDataInstance.choices.put({
+              id: 'enforceOnlyOneRecordEverStoredAndOverwritten',
+              choices: { ...apiData.choices?.updates, uiState_pushToApi: false },
+            })
+          }
+          if (apiDataType !== 'choices') {
+            const updates = apiData[apiDataType]?.updates ?? []
+            const updatesWithPushToApiTagReset = resetPushToApiTagFromItems(updates)
+            const deletes = apiData[apiDataType]?.deletes ?? []
+            const removes = apiData[apiDataType]?.removes ?? []
+            const deleteIds = deletes.map(({ id }) => id)
+            const removeIds = removes.map(({ id }) => id)
+            const error = apiData[apiDataType]?.error
+            const is401Or403 = error && [401, 403].includes(error.code)
+            const is403 = error?.code === 403
```

Comment:
Added the use of `removes`, `removeIds`, and `is403`

Reply:
nice!

Reply:
a bit confusing having both` is401Or403 `and `is403`. Would probably just keep both separate if you're not always using them together

Reply:
good call